src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java

```diff
@@ -600,6 +600,11 @@ protected Query doToQuery(QueryShardContext context) {
             throw new IllegalArgumentException(String.format(Locale.ROOT, "Engine [%s] does not support filters", knnEngine));
         }
 
+        // rewrite filter query if it exists to avoid runtime errors in next steps of query phase
+        if (Objects.nonNull(filter)) {
```

Comment:
Should we be rewriting the filter query in doRewrite method of querybuilder? A caveat of that is we have to replace the filter with rewritten one, which we are doing here as well. Just asking because there is dedicated method to rewrite querybuillders

Reply:
> Should we be rewriting the filter query in doRewrite method of querybuilder?

+1 on this. I am also little surprised why we are doing the rewrite in doToQuery function.

Reply:
other queries do rewrite as part of `doToQuery`, it's even part of the abstract [query builder class](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractQueryBuilder.java#L250). I don't have strong preference, just we already have override for `doToQuery`, but not for `doRewrite`. 
Let me check if there are any caveats, if not I can move the logic to `doRewrite`

Reply:
great.

Reply:
> it's even part of the abstract [query builder class](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractQueryBuilder.java#L250).

This seems to be the actual execution of OS, where `.rewrite` calls doRewrite of underlying querybuilder and `toQuery` will call doToQuery. They are chained but called one after the other

Reply:
yup, I checked, override in `rewrite` does the job, pushed a commit that implements the change